### ISSUE TITLE
Allow resolving merge conflicts "interactively"

### ISSUE THREAD

**Issue by SlugFiller** [gitextensions/gitextensions#11346]
### Feature description

When performing an interactive rebase, if one of the commits is set to "edit", the rebase is paused, the dialog is hidden, you're allowed to freely browse the repository, files, different revisions, make any changes, stage the changes you want, close and restart the client, and only when you're ready to continue, you can simply press a button at the top to accept the changes, create the commit, and continue with the rebase process.

If, on the other hand, you encounter a merge conflict, regardless of whether it's a merge, rebase, or interactive rebase, you get stuck in a tiny modal dialog that asks you to merge the files one by one, does not allow you to look at other files, let alone other revisions, does not allow you to make any other edits, does not allow you to re-do an edit on a file that's already "merged", and is generally extremely limiting.

While I get that having access to 3-way diffs is a very special condition, that is not easily represented in the existing default UI. But it doesn't have to be as hostile to the user as the current conflict resolution dialog. 

An example where this is particularly problematic, is if a function is moved from one file to another, and a commit that changes that function is rebased on top of the move. A file-by-file conflict resolution requires the change to be deleted. But a proper resolution requires the change to be reapplied to the function's new location. Not only does the conflict resolution dialog not allow this, but if it's a multi-commit rebase, it will chug along, adding multiple commits on top of a commit that must be amended for the rebase to be proper.

Also, having to choose if to keep or delete a file without being able to see its contents is super problematic.

### Environment

- Git Extensions 4.1.0.16698
- Build 1fe52a137acef1fbd2b653dea263b44ad5844177
- Git 2.42.0.windows.1
- Microsoft Windows NT 10.0.19045.0
- .NET 6.0.24
- DPI 96dpi (no scaling)
- Portable: False
- Microsoft.WindowsDesktop.App Versions

```
    Microsoft.WindowsDesktop.App 6.0.21 [D:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
    Microsoft.WindowsDesktop.App 6.0.24 [D:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
```

**Comment by mstv** [comment]
I do not like the merge dialog very much either. Though it has a context menu and both dialogs can be closed and re-opened, i.e. you are not locked in.

(More information about remote / local and theirs / ours would be very helpful - as discussed in a few former issues like #11318 - particularly because they interchange roles dependent on whether merging or rebasing. We could reuse the function `DescribeRevision` as used for the groups in the FileStatusList - if the related commits can easily be determined.)

![image](https://github.com/gitextensions/gitextensions/assets/36601201/f294b887-4648-41ed-b4fe-9a3a206d9008)

What would you add and how?

**Comment by RussKie** [comment]
FWIW, I have few POCs in this area.

1. Load the conflict data in async way, making the dialog responsive

2. Provide the ability to specify merge strategies (here and in the cherry
pick dialog).

**Comment by SlugFiller** [comment]
Ah, you're right. Having to press the "X" is a bit confusing. So, first off, I would add a button next to "Abort". Not sure what I would call it. "Manually review"? "Pause"? Currently, the logical assumption is that "X" and "Abort" do the same thing. But at least it's only a UX issue.

The "open local/remote/base with" options have the same issues as their diff list and file tree equivalents. i.e. I will probably need to make a PR adding `IScriptOptionsProvider` to this dialog.

Other than that it would be nice if the diff list and view directly had a way to treat a file as being in merge conflict. That means:
- Having the same right click options as in the resolve window. At the very least, "Mark as resolved" and "Open with mergetool". Although I would personally add all of them.
- Have an option to show Ours->Base and Theirs->Base in the diff view, instead of just Conflict->Base. Maybe something like how merge commits are shown, but instead of `Diff with A/B` it would have `Diff from A/B`.
- Have per-version `Save Selected As...`. This could be done with new options. But a better UX would be to be able to do this from the `Diff from A/B` versions.
- If possible, mark the two commits in some way, and/or allow to quickly navigate to them. One commit is available as the parent of the index, but the other is not.
- Actually, just adding the second parent as a commit index parent might be good enough, and it would also allow having a `Diff with B` to see how the commit index compares to the rebase source and not just to the rebase target.
- Working directory should also have the commit index's parent/s as a parent. And not just during rebase/merge. To be able to see the total diff, rather than just diff to the index.

I did notice that merging asks you if you want to resolve the conflicts, while rebase does not. It's probably because rebase has a multi-commit dialog. But it would still be nice to have some uniformity. Actually, I would personally rather it didn't ask, and immediately assumed "No". The big `Resolve...` button at the top already serves the same purpose of the popup, in roughly the same screen space, so it's just duplicate UI.

Also, rebasing a deleted file on top of a modified file causes some very weird results. The working directory is showing the file deleted, while the commit index is showing it as the modified file. But the actual status of the working directory is that the file still exists. Doing the opposite operation - modified over delete, shows the index as deleted, and working directory as modified, with the file actually being in the working directory. In both cases, marking the conflict as "resolved" keeps the file, with the file being added to the index in the latter case, and being gone from both index and working directory in the prior case. Kinda feels like a bug, but I dunno.

Oof. Ended up having a lot to say. Ended up a bit large for a single feature request. But it's all interconnected.

**Comment by SlugFiller** [comment]
Slightly alternative approach:
1. Add all right click options from `Resolve` dialog to index/working directory right click menu for files in conflict
2. Remove the `Resolve` dialog completely
3. In `Rebase` dialog, clicking `>Solve Conflicts<` simply closes the dialog.
4. So long as a conflict exists, the only options at the top are `Abort` and (for rebase only) `More...`. The `Resolve...` option is removed.
5. Once all conflicts are marked as resolved, the `Continue` option appears.

It's simple, and is way better UX.

Side note: New icons need to be added for "Added and merge conflict" and "Deleted and merge conflict". Especially the prior, since the latter at least appears as merge conflict in the index, while the prior just appears as added in both index and working directory.

**Comment by mstv** [comment]
> Having to press the "X" is a bit confusing.

I agree, the similarity to "Abort" may confuse users. Perhaps we should reopen the discussion towards "Abort merge" & "Abort rebase" again.
Anyway, there is no need to click "X", I prefer pressing `Esc`. ;)

> 1. Add all right click options from Resolve dialog to index/working directory right click menu for files in conflict

It would be good to wait with this until the FileStatusList has been refactored (and switched to TreeView).
We should not maintain the (same) code twice: in FormBrowse and in FormCommit. The context menus should be same everywhere.

> But it would still be nice to have some uniformity.

👍

> Actually, I would personally rather it didn't ask

Does this option improve the UX for you?

![image](https://github.com/gitextensions/gitextensions/assets/36601201/21b21e68-a30f-47bc-a818-74ad1ee074e5)

**Comment by SlugFiller** [comment]
> Does this option improve the UX for you?

Despite the title, it has no influence on rebase.

For merge, it does the opposite of what I want: It defaults to "Yes" (open the dialog), when I want it to default to "No" (Show me the repository with conflicts, and let me choose if and when I want to open the dialog)

**Comment by SlugFiller** [comment]
Having gone through another rebase, I found another thing I want: After merging a file, do not auto-stage. Unstaged files are the fastest way to see a file that was in conflict. And staging the file makes it harder to keep track of which files had conflicts, in case there was a conflict in which requires reading or merging a different file gives more hints as to how to resolve it.

Of course, not blocking access to the repo view during merge in the first place would also make this less of a pain. Just let KDiff3 run in the background, and let the user stage the final file when it's ready. Created temp files can be deleted when clicking "Continue".
